pkg/operator/controllers/machine/machine_controller_test.go

```diff
@@ -0,0 +1,322 @@
+"image": {
+"publisher": "azureopenshift",
+"offer": "aro4"
+},
+"vmSize": "Standard_D4s_v3"
+}`)},
+				},
+			},
+		}
+
+		workerMachineSet := &machinev1beta1.MachineSet{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "workermachineset",
+				Namespace: machineSetsNamespace,
+			},
+			Spec: machinev1beta1.MachineSetSpec{
+				Replicas: to.Int32Ptr(3),
+			},
+		}
+		return maofake.NewSimpleClientset(worker0, worker1, worker2, master0, master1, master2, workerMachineSet)
+	}
+
+	// define fake cluster resource
+	newFakeAro := func(a *arov1alpha1.Cluster) *arofake.Clientset {
+		return arofake.NewSimpleClientset(a)
+	}
+
+	baseCluster := arov1alpha1.Cluster{
+		ObjectMeta: metav1.ObjectMeta{Name: "cluster"},
+		Status:     arov1alpha1.ClusterStatus{Conditions: []status.Condition{}},
+	}
+
+	tests := []struct {
+		name           string
+		request        ctrl.Request
+		maocli         *maofake.Clientset
+		arocli         *arofake.Clientset
+		wantConditions []status.Condition
+		wantErr        bool
+	}{
+		{
+			name:   "valid",
+			maocli: newFakeMao("512", "azureopenshift", "Standard_D4s_v3"),
+			arocli: newFakeAro(&baseCluster),
+			wantConditions: []status.Condition{{
+				Type:    arov1alpha1.MachineValid,
+				Status:  corev1.ConditionTrue,
+				Message: "All machines valid",
+				Reason:  "CheckDone",
+			}},
+		},
+		{
+			name:   "wrong_vm_size",
+			maocli: newFakeMao("512", "azureopenshift", "Standard_D4s_v9"),
+			arocli: newFakeAro(&baseCluster),
+			wantConditions: []status.Condition{{
+				Type:    arov1alpha1.MachineValid,
+				Status:  corev1.ConditionFalse,
+				Message: "machine foo-hx8z7-worker-0: invalid VM size 'Standard_D4s_v9'",
+				Reason:  "CheckFailed",
+			}},
+		},
+		{
+			name:   "wrong_disk_size",
+			maocli: newFakeMao("64", "azureopenshift", "Standard_D4s_v3"),
+			arocli: newFakeAro(&baseCluster),
+			wantConditions: []status.Condition{{
+				Type:    arov1alpha1.MachineValid,
+				Status:  corev1.ConditionFalse,
+				Message: "machine foo-hx8z7-worker-0: invalid disk size '64'",
+				Reason:  "CheckFailed",
+			}},
+		},
+		{
+			name:   "wrong_image_publisher",
+			maocli: newFakeMao("512", "bananas", "Standard_D4s_v3"),
+			arocli: newFakeAro(&baseCluster),
+			wantConditions: []status.Condition{{
+				Type:    arov1alpha1.MachineValid,
+				Status:  corev1.ConditionFalse,
+				Message: "machine foo-hx8z7-worker-0: invalid image '{bananas aro4   }'",
+				Reason:  "CheckFailed",
+			}},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			r := &MachineReconciler{
+				maocli:                 tt.maocli,
+				log:                    logrus.NewEntry(logrus.StandardLogger()),
+				arocli:                 tt.arocli,
+				isLocalDevelopmentMode: false,
+				role:                   "master",
+			}
+
+			_, err := r.Reconcile(context.Background(), tt.request)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("Machine.Reconcile() error = %v, wantErr %v", err, tt.wantErr)
+				return
```

Comment:
1. `t.Fatal`/`t.Fatalf` is going to stop subtest execution.
2. We just tend to print an actual error. If the test fails - it will give you the test name + line number from which you can find the desired test result.

```suggestion
				t.Fatalf(err)
```

Reply:
Also. It looks like you actually do not use `wantErr`. I think in this case it is ok not to test the error which `Reconcile` returns beucase the only error it returns comes from `controllers.SetCondition(...)` which should have a test own test. In master it currently has no test, but https://github.com/Azure/ARO-RP/pull/1508 add coverage.

So we can remove `wantErr`.